base/inference.jl

```diff
@@ -1221,16 +1232,58 @@ function findlabel(labels, l)
 end
 
 function label_counter(body)
-    l = 0
+    l = -1
     for b in body
         if isa(b,LabelNode) && (b::LabelNode).label > l
             l = (b::LabelNode).label
         end
     end
-    l
+    return l
 end
 genlabel(sv) = LabelNode(sv.label_counter += 1)
 
+function find_gensym_uses(body)
+    uses = Vector{Int}[]
+    for line = 1:length(body)
+        find_gensym_uses(body[line], uses, line)
+    end
+    return uses
+end
+function find_gensym_uses(e::ANY, uses, line)
+    if isa(e,GenSym)
+        id = (e::GenSym).id+1
+        while length(uses) < id
+            push!(uses, Array(Int,0))
+        end
+        push!(uses[id], line)
+    elseif isa(e,Expr)
+        b = e::Expr
+        head = b.head
+        if head === :line
+            return
+        end
+        if head === :(=)
+            if isa(b.args[1],GenSym)
+                id = (b.args[1]::GenSym).id+1
+                while length(uses) < id
+                    push!(uses, Array(Int,0))
+                end
+            end
+            find_gensym_uses(b.args[2], uses, line)
+            return
+        end
+        for a in b.args
+            find_gensym_uses(a, uses, line)
+        end
+    end
+end
+
+function GenSym(sv::StaticVarInfo, typ)
```

Comment:
I don't think it's good style to have a constructor mutate something. I'd call this something like `newvar!(sv, typ)`.